### PR TITLE
Rhmap 7365

### PR DESCRIPTION
**Pull request by philbrookes**
- Rebased to master (89 commits).
- Updated mongo driver to 2.1.11
- Made compatible with Mongo 2.4

My changes are in this commit:
https://github.com/fheng/agenda/commit/3c0b2fb39a11d2d95ba999f83d315b4f5d56ca15

### CODE REVIEW

**Comment by philbrookes** [comment]
Various test cases can be seen passing in the upstream repos Jenkins server:
https://travis-ci.org/rschmukler/agenda/builds/135904456

File: lib/agenda.js [hunk 1]
```diff
@@ -94,17 +96,42 @@ Agenda.prototype.database = function(url, collection, options, cb) {
...
+                                function (err, result) {
+                                  try {
+                                    if (err) {
+                                      self.emit('error', err);
+                                    } else {
+                                      self.emit('ready');
+                                    }
+                                    if (cb) cb(err, self._collection);
+                                  } catch(e){
+                                    if(e.message === 'no such cmd: createIndexes'){
```
Comment (luigizuccarelli) [inline lib/agenda.js]: @philbrookes - is this perhaps too strict a test - just a thought ?

Comment (philbrookes) [inline lib/agenda.js]: I'm not sure, I wanted to be strict, to ensure I didn't accidentally try to handle an exception caused by a different problem. The problem was that there doesn't seem to be an exception "code":
```
{ 
  [MongoError: no such cmd: createIndexes]
  name: 'MongoError',
  message: 'no such cmd: createIndexes',
  ok: 0,
  errmsg: 'no such cmd: createIndexes',
  'bad cmd': { createIndexes: 'agendaJobs', indexes: [ [Object], [Object] ] } 
}
```

File: lib/agenda.js [hunk 2]
```diff
@@ -403,32 +487,136 @@ function processJobs(extraJob) {
     for (jobName in definitions) {
       jobQueueFilling(jobName);
     }
-  } else {
-    // On the fly lock a job
+  } else if (definitions[extraJob.attrs.name]) {
```
Comment (luigizuccarelli) [inline lib/agenda.js]: @philbrookes - were these updates from a previous version or did you add them L490 - L601 ?

**Comment by luigizuccarelli** [comment]
Executed npm test 

| MongoDB || Node0.10 || Node4.4 |
| 2.4.6 || Pass || Pass |
|  3.2.5 || Pass || Pass |

File: lib/agenda.js (hunk 2 above)
Comment (philbrookes) [inline lib/agenda.js]: @luigizuccarelli These came from the rebase.

**Comment by luigizuccarelli** [comment]
@philbrookes - 👍  - LGTM (was going to ask you about the tests - but that seems to be also part of the rebase)

**Pull request merged**